File: binary_search.py
def binary_search(data, target):
  """
  This function implements a binary search algorithm.

  Args:
      data: A sorted list of numbers to search through.
      target: The number to search for.

  Returns:
      The index of the target number in the list if found, otherwise -1.
  """

  low = 0
  high = len(data) - 1

  while low <= high:
    mid = (low + high) // 2

    if data[mid] == target:
      return mid
    elif data[mid] < target:
      low = mid + 1
    else:
      high = mid - 1

  return -1

# Example usage (assuming the data list is already sorted)
data = [1, 5, 8, 9, 12]
target = 8

result = binary_search(data, target)

if result != -1:
  print(f"The target number {target} was found at index {result}")
else:
  print(f"The target number {target} was not found in the list")

File: BinarySearch.java
public class BinarySearch {

  public static int binarySearch(int[] arr, int target) {
    int low = 0;
    int high = arr.length - 1;

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (arr[mid] == target) {
        return mid;  
      } else if (arr[mid] < target) {
        low = mid + 1;  
      } else {
        high = mid - 1; 
      }
    }
    return -1; 
  }

  public static void main(String[] args) {
    int[] arr = {2, 5, 8, 12, 16};
    int target = 12;

    int result = binarySearch(arr, target);

    if (result == -1) {
      System.out.println("Target not found in the array.");
    } else {
      System.out.println("Target found at index: " + result);
    }
  }
}
